### ISSUE TITLE
Much better wav2vec2 performance  after using the new version of fairseq?

### ISSUE THREAD

**Issue by Sanyuan-Chen** [s3prl/s3prl#127]
Hi, Thanks for the great open-source toolkit! 

I noticed you updated the installation note and asked to install a new version of fairseq due to some known bugs for wav2vec2 inference and ASR decoding.

About a month ago, I tried evaluated wav2vec2 with the older version of your repo and the stable version of fairseq, and I got similar results to the results you reported.

However, After using the new version of fairseq, and re-evaluate wav2vec2 on some downstream tasks, I find it can achieve 77.31 on SID task (lr=1e-1) and 6.81 on SD task  (lr=1e-2), which is much better than the reported results (45.62 on SID and 7.48 on SD).

It seems very surprising to me, and I'm wondering if you have the results after using the new version of fairseq? Are we supposed to get a better performance? 

BTW, can you please provide a reference of the "known bugs for wav2vec2 inference and ASR decoding" in the stable version 0.10.2 of fairseq?

Many thanks~

**Comment by leo19941227** [comment]
Hi,

Thanks you for the interest!
May I ask which commit you are currently using in this repo?
We had a major update to let the downstream training weighted-sum on all the hidden states by default 6 days before in this commit: https://github.com/s3prl/s3prl/commit/39b408730bb6b39b6d10d963b3eba2b82582e735

Since we found that especially for wav2vec2, the different layers have huge performance differences on different tasks, and it will be more fair the use weighted-sum to represent the best a pretrained model can get. So what you got is very likely the weighted-sum results, and the numbers are also closed to what we got here. (We have also updated the paper to use weighted-sum and the Arxiv version will be available a few days later.)

The fairseq bugs we consider in this repo are fixed by these two commits: [fix asr decoding](https://github.com/pytorch/fairseq/commit/cb84694c195afced474d17318b5e746d1a9d20a3) and [fix conv padding](https://github.com/pytorch/fairseq/commit/4fed0beca64a52aa718371dc3b2cf1fd979197a4).
The first fix is required for the CTC decoder to function normally, while the second fix is pretty much a correctness issue.

Hope these information help!

Sincerely,
Leo

**Comment by Sanyuan-Chen** [comment]
Thank you for the response, It's really helpful!  
Yes, I am currently using the commit after the weighted-sum update. It is good to know the reason for the performance improvement and the huge performance differences among different hidden states.
Looking forward to your updated superb paper!

**Comment by Sanyuan-Chen** [comment]
Hi Leo, 

Thanks again for the response above, and I can successfully get similar results on almost all the downstream tasks as reported in your paper.

However, I am wondering if you are also using the weighted-sum hidden states when you report the results on QBE tasks? 
When I try to experiment with the Hubert-base model, I got the score **0.0579** with **the weighted-sum representation** and **0.0695** with **the last layer's representation**, while the **reported score** is **0.0736**.
When I try to experiment with the Hubert-large model, I got the score **0.0176** with **the weighted-sum representation** and **0.0339** with **the last layer's representation**, while the **reported score** is **0.0353**.

Since the results I got are not exactly the same as your reported results, I am wondering if you can offer any insights?

Thanks a lot!

**Comment by leo19941227** [comment]
Hi Sanyuan-Chen,

Since QbE actually does not involve downstream model training, so you results is very likely the **average** of all layers, since our initial values of those **weights** are uniformly distributed. In the paper, we run DTW for all the hidden states one-by-one. Since DTW is less computationally expensive than downstream model training so layer exploration is feasible. To do this, instead of using the default `-s hidden_states`, you can use `-s hidden_state_0`, `-s hidden_state_1` until `-s hidden_state_12` for HuBERT base. You can also automate this pipeline with a shell script. Hence, our newly reported results should be always better than the last layer representation.

I should also update the downstream README to clarify this part for QbE.
Many thanks for letting me know!

**Comment by leo19941227** [comment]
Hi @Sanyuan-Chen,

Glad to know it is reproducible at your side!
